### PR TITLE
feat(load) delay BufRead if inside packer_compile

### LINKED ISSUE #778: Execute BufRead later
When `packer` sees that a plugin requires `BufRead` (ie by detecting eg a `ftdetect` folder) it does `doautocmd BufRead`.

This happens sometimes too early. For example I had an issue where LSPs didn't start automatically when opening a file using nvim (by doing `nvim some_file`). After a very long time of debugging and trying do understand what was going on, I realised that the following was happening:

In `packer_compiled` when calling all the `setup`/`config` for various plugins the `config` for `nvim-gps` was called before `lspconfig`. `nvim-gps` in turn `require` `nvim-treesitter` by module and I have `nvim-treesitter` lazy loaded by module. So `packer` starts loading `nvim-treesitter` which has a `ftdetect` folder which makes `packer` trigger `BufRead`. However all this happens before `lspconfig` is loaded and the `FileType` autocmds that `lspconfig` adds are therefore never executed.

I think that the better approach would maybe be to:
* keep track of when `packer_compiled` is entered/exited.
* if `needs_bufread` is true while executing `packer_compiled`, don't trigger `BufRead`.
* if by the end of `packer_compiled` at least one of the plugins `needs_bufread` then trigger it.

### PR DESCRIPTION
**Pull request by AckslD**
Closes #778

### CODE REVIEW

File: lua/packer/compile.lua [hunk 1]
```diff
@@ -33,6 +33,20 @@ vim.api.nvim_command('packadd packer.nvim')
 local no_errors, error_msg = pcall(function()
 ]]
 
+
+local enter_packer_compile = [[
+vim.g.packer_inside_compile = true
```
Comment (wbthomason) [inline lua/packer/compile.lua:38]: Nitpick: I'd prefer that this/the other `vim.g` variables currently used be switched to the `_G.packer` table for consistency.

**Review by wbthomason — CHANGES_REQUESTED** [review]
LGTM, except for requesting a change in global scope convention.

File: lua/packer/compile.lua (hunk 1 above)
Comment (AckslD) [reply lua/packer/compile.lua:38]: Sure thing, I tried to grep for this but only found `_G.packer_plugins`, is this the one you mean? Or did you mean a new table?

Comment (wbthomason) [reply lua/packer/compile.lua:38]: We create a `_G._packer` in e.g. line 76 of packer.lua or line 94 of compile.lua - this is meant to be the global "packer metadata" table. It's just slightly nicer than relying on variables with specific names in Vim's global scope, imo.

Comment (AckslD) [reply lua/packer/compile.lua]: Ah I see, I missed the `_` prefix of `packer`. Done in 5effb94da662e5052cd1ec3d958853efb57bcd90 :+1:

**Review by wbthomason — APPROVED** [review]
LGTM, thank you!

**Comment by wbthomason** [comment]
Check failed with an unrelated error: `gzip: stdin: not in gzip format`

**Pull request merged**